src/funnel.test.ts

```diff
@@ -0,0 +1,1049 @@
+      // after that is simply dropped.
+      expect(mockFn).toHaveBeenLastCalledWith(["d"]);
+    });
+
+    test("invokedAt: both", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        invokedAt: "both",
+        delayMs: UT,
+      });
+      foo.call("a");
+      foo.call("b");
+      foo.call("c");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+      await sleep(UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+      expect(mockFn).toHaveBeenLastCalledWith(["b", "c"]);
+
+      foo.call("d");
+      foo.call("e");
+      foo.call("f");
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+
+      await sleep(UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(3);
+      expect(mockFn).toHaveBeenLastCalledWith(["d", "e", "f"]);
+    });
+
+    it("never fires when only delay is used and invoked only at end", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        invokedAt: "end",
+        delayMs: UT,
+      });
+      foo.call("a");
+      foo.call("b");
+      foo.call("c");
+
+      expect(mockFn).toHaveBeenCalledTimes(0);
+
+      await sleep(2 * UT);
+      foo.call("d");
+      foo.call("e");
+      foo.call("f");
+
+      expect(mockFn).toHaveBeenCalledTimes(0);
+
+      await sleep(2 * UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(0);
+    });
+
+    test("invocations in the middle of a window", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        invokedAt: "both",
+        delayMs: UT,
+      });
+      foo.call("a");
+      foo.call("b");
+      foo.call("c");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+      await sleep(0.75 * UT);
+      foo.call("d");
+      foo.call("e");
+      foo.call("f");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+
+      await sleep(0.75 * UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+      expect(mockFn).toHaveBeenLastCalledWith(["b", "c", "d", "e", "f"]);
+    });
+  });
+
+  describe("burst timer", () => {
+    test("invokedAt: start", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        invokedAt: "start",
+        burstCoolDownMs: UT,
+      });
+      foo.call("a");
+      foo.call("b");
+      foo.call("c");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+      await sleep(UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+
+      foo.call("d");
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+      // "b" and "c" have been ignored because the funnel is only configured to
+      // run the function at the start of the delay window, meaning anything
+      // after that is simply dropped.
```

Comment:
**question:** do you think this behavior is desirable?

i think it is... but i'm not sure.

Reply:
There is no other logical way to handle them. If execution is set to "start" it means there is no later timing they could be used at.